File: array/prefix_sum/advanced/0560_subarray_sum_equals_k.js
/*
Given an array of integers nums and an integer k, return the total number of 
subarrays whose sum equals to k.

A subarray is a contiguous non-empty sequence of elements within an array.
*/

// https://leetcode.com/problems/subarray-sum-equals-k/description/
// related https://leetcode.com/problems/binary-subarrays-with-sum/description/

// a n-length array has n*(n+1)/2 subarrays
// brute force requires O(n^2) time complexity
// use prefix sum map to reduce time complexity to O(n)
var subarraySum = function (nums, goal) {
  let cnt = new Map(),
    preSum = 0,
    ans = 0;
  // loop through the array
  for (let i = 0; i < nums.length; i++) {
    // count the times the same preSum appears from nums[0] to nums[n-1]
    cnt.set(preSum, (cnt.get(preSum) || 0) + 1);
    // update the prefix sum
    preSum += nums[i];
    // if the sum of nums[i] to nums[j] equals to goal,
    // then preSum[j] - preSum[i] = goal => preSum[i] = preSum[j] - goal
    // so preSum[i] is the key to find the number of subarrays that satisfy the condition
    // cnt.get(preSum - goal) is the number of subarrays that satisfy the condition
    ans += cnt.get(preSum - goal) || 0;
  }
  return ans;
};

File: array/prefix_sum/advanced/0560_subarray_sum_equals_k.py
# Given an array of integers nums and an integer k, return the total number of subarrays whose sum equals to k.

# A subarray is a contiguous non-empty sequence of elements within an array.

# https://leetcode.com/problems/subarray-sum-equals-k/
# same problem LC 930 https://leetcode.com/problems/binary-subarrays-with-sum/

class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        preSum = res = 0
        cnt = {0:1}  # Initialize with 0 having count 1
        
        for num in nums:
            preSum += num 
            res += cnt.get(preSum - k, 0)
            cnt[preSum] = cnt.get(preSum, 0) + 1
        
        return res


class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        preSum = res = 0
        cnt = dict()
        for num in nums:
            cnt[preSum] = cnt.get(preSum, 0) + 1
            preSum += num
            # 如果存在某个子数组和为k, cnt[preSum - k]即是满足条件的子数组的数量
            # 假设目前循环到nums[j], 则nums[:j]之和为preSum, 如果存在某个点i < j，使得nums[i:j]之和为preSum - k，那么nums[:i]之和为k。
            # cnt[preSum - k]返回符合条件的 i 的个数
            res += cnt.get(preSum - k, 0)
        return res